bin/varnishd/cache/cache_hash.c

```diff
@@ -1143,12 +1144,11 @@ HSH_DerefObjCore(struct worker *wrk, struct objcore **ocp)
 	r = --oc->refcnt;
 	if (!r)
 		VTAILQ_REMOVE(&oh->objcs, oc, hsh_list);
-	hsh_rush1(wrk, oc, &rush);
 	Lck_Unlock(&oh->mtx);
-	hsh_rush2(wrk, &rush);
 	if (r != 0)
 		return (r);
 
+	AZ(oc->flags & OC_F_BUSY);
```

Comment:
By now, you have really lost me with the reinterpretation of `OC_F_BUSY`. I hope I understand what you are trying to achieve (in fact, I was the person originally reducing spurious rushes), and I support the goal, but I really can not follow why we need all these fundamental changes and they do not look right to me. If anything, it seems I need more explanations. Yes, I did read the commit messages and PR description.

Reply:
The removal of the busy flag for synth/req_body objcores and the `HSH_Fail()` and `HSH_{Abandon,Withdraw}()` changes are here to ensure that by the time we end a fetch task or decide not to engage a fetch the objcore is no longer considered busy.

Before this pull request, this is only the case for the happy path via `HSH_Unbusy()` and it forces all kinds of fallbacks leading to spurious waiting list wake ups.

Making sure all paths and not just the happy paths clear the busy flag means that the safety nets can go away and be replaced by assertions that the flag is indeed cleared.

Reply:
Marking this one as resolved again, I believe this is also covered by the https://github.com/varnishcache/varnish-cache/pull/4032#discussion_r1440326251 thread.